source/memcached-store.ts

```diff
@@ -0,0 +1,171 @@
+// /source/memcached-store.ts
+// A `memcached` store for the `express-rate-limit` middleware.
+
+import { promisify } from 'node:util'
+import type {
+	Store,
+	Options as RateLimitConfiguration,
+	IncrementResponse,
+} from 'express-rate-limit'
+import Memcached from 'memcached'
+import type { Options, MemcachedClient } from './types'
+
+/**
+ * A `Store` for the `express-rate-limit` package that stores hit counts in
+ * Memcached.
+ */
+class MemcachedStore implements Store {
+	/**
+	 * The number of milliseconds to remember that user's requests.
+	 */
+	expiration!: number
+
+	/**
+	 * The text to prepend to the key.
+	 */
+	prefix!: string
+
+	/**
+	 * The `memcached` client to use.
+	 */
+	client!: MemcachedClient
+
+	/**
+	 * @constructor for `MemcachedStore`.
+	 *
+	 * @param options {Options} - The options used to configure the store's behaviour.
+	 */
+	constructor(options?: Partial<Options>) {
+		this.prefix = options?.prefix ?? 'rl:'
+
+		if (options?.client) {
+			if (
+				typeof options.client.get === 'function' &&
+				typeof options.client.set === 'function' &&
+				typeof options.client.del === 'function' &&
+				typeof options.client.incr === 'function' &&
+				typeof options.client.decr === 'function'
+			)
+				this.client = options.client
+			else throw new Error('An invalid memcached client was passed to store.')
+		} else
+			this.client = new Memcached(
+				options?.locations ?? ['localhost:11211'],
+				options?.config ?? {},
+			)
+	}
+
+	/**
+	 * Method that actually initializes the store.
+	 *
+	 * @param options {RateLimitConfiguration} - The options used to setup the middleware.
+	 *
+	 * @impl
+	 */
+	init(options: RateLimitConfiguration) {
+		this.expiration = options.windowMs / 1000 // [in seconds]
+	}
+
+	/**
+	 * Method to prefix the keys with the given text.
+	 *
+	 * @param key {string} - The key.
+	 *
+	 * @returns {string} - The text + the key.
+	 */
+	prefixKey(key: string): string {
+		return `${this.prefix}${key}`
+	}
+
+	/**
+	 * Method to suffix the keys with `__expiry`, which is the name of the key used
+	 * to store the reset timestamp for that key.
+	 *
+	 * @param key {string} - The key.
+	 *
+	 * @returns {string} - The key + '__expiry'.
+	 */
+	expirySuffix(key: string): string {
+		return `${key}__expiry`
+	}
```

Comment:
Oh, I probably should have said to add expiry to the prefix rather than the suffix, e.g. 
```suggestion
	expiryKey(key: string): string {
		return `${this.prefix}expiry:${key}`
	}
```

And then pass in the original key rather than the prefixed one.

It's not a big deal, but this feels a little cleaner.

Reply:
Done in commit 3899485.